### PR TITLE
[Core] Tmp_dir is incorrectly set from the driver when `ray start --temp-dir` is given

### LINKED ISSUE #39385: [Core] _node_ip_address causing issue when head node started at a temp dir
### What happened + What you expected to happen

This PR https://github.com/ray-project/ray/pull/37644 introduced a bug that if you start a ray cluster by specifying a temp-dir that is not the default temp dir, then when you call `ray.init(address)` it still tries to read the `node_ip_address.json` file from default temp dir.
The root cause is that `ray.init(address)` doesn't contain `temp_dir`, and inside node creation,  it updates ray_params.temp_dir if it's absent (https://github.com/ray-project/ray/blob/8b7fcd7bb8341cf2431d431a55cd07ad4b74d184/python/ray/_private/node.py#L112-L126) so that it defaults to a different temp dir than the head node's temp dir.

### Versions / Dependencies

Latest nightly
https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-3.0.0.dev0-cp310-cp310-manylinux2014_x86_64.whl 

### Reproduction script

Use `ray start --head --temp-dir '/local_disk0/tmp/ray-some-temp-dir'` to start the head node

```
import ray
ray.init(address=address_of_the_head_node)
```

It will show up the stacktrace like:
```
2023-09-07 13:55:59,011	INFO worker.py:1458 -- Connecting to existing Ray cluster at address: 10.68.173.106:6379...
2023-09-07 13:56:00,019	INFO node.py:996 -- Can't find a `node_ip_address.json` file from /tmp/ray/session_2023-09-07_13-55-41_401531_2199. Have you started Ray instsance using `ray start` or `ray.init`?
2023-09-07 13:56:10,031	INFO node.py:996 -- Can't find a `node_ip_address.json` file from /tmp/ray/session_2023-09-07_13-55-41_401531_2199. Have you started Ray instsance using `ray start` or `ray.init`?
2023-09-07 13:56:20,043	INFO node.py:996 -- Can't find a `node_ip_address.json` file from /tmp/ray/session_2023-09-07_13-55-41_401531_2199. Have you started Ray instsance using `ray start` or `ray.init`?
2023-09-07 13:56:30,054	INFO node.py:996 -- Can't find a `node_ip_address.json` file from /tmp/ray/session_2023-09-07_13-55-41_401531_2199. Have you started Ray instsance using `ray start` or `ray.init`?
2023-09-07 13:56:40,067	INFO node.py:996 -- Can't find a `node_ip_address.json` file from /tmp/ray/session_2023-09-07_13-55-41_401531_2199. Have you started Ray instsance using `ray start` or `ray.init`?
2023-09-07 13:56:50,078	INFO node.py:996 -- Can't find a `node_ip_address.json` file from /tmp/ray/session_2023-09-07_13-55-41_401531_2199. Have you started Ray instsance using `ray start` or `ray.init`?
ValueError: Can't find a `node_ip_address.json` file from /tmp/ray/session_2023-09-07_13-55-41_401531_2199. for 60 seconds. A ray instance hasn't started. Did you do `ray start` or `ray.init` on this host?
```

### Issue Severity

High: It blocks me from completing my task.

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

Looks like the temp dir was never set correctly when ray.init is called after it is set from ray start.

It is because we always eagerly set the temp dir before we obtain it from GCS (via internal_kv_get("temp_dir")).

This PR defers the temp_dir detection to the later part of the code to fix the issue

## Related issue number

Closes https://github.com/ray-project/ray/issues/39385

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rkooo567** [comment]
seems like the master is unhappy...

**Comment by WeichenXu123** [comment]
Thanks for addressing this! :)

File: python/ray/tests/test_ray_init_2.py [hunk 1]
```diff
@@ -376,6 +378,20 @@ def test_driver_node_ip_address_auto_configuration(monkeypatch, ray_start_cluste
...
 
 
+@pytest.fixture
+def short_tmp_path():
+    path = tempfile.mkdtemp(dir="/tmp")
+    yield path
+    shutil.rmtree(path)
+
+
+def test_temp_dir_with_node_ip_address(ray_start_cluster, short_tmp_path):
```
Comment (vitsai) [inline python/ray/tests/test_ray_init_2.py:388]: Should we also test the `init(_temp_dir=x)` path, by initing first, then adding a node and checking if that node has the same temp_dir?

Comment (rkooo567) [reply python/ray/tests/test_ray_init_2.py:388]: I think adding a node to the cluster thats' started by ray.init is kind of weird pattern (I've never seen anyone doing this), so it is probably okay.

**Review by fishbone — APPROVED** [review]

**Comment by rkooo567** [comment]
Hmm looks like spark on ray somehow failed (unexpected). Investigating..

**Comment by rkooo567** [comment]
@WeichenXu123 do you have any idea? Not very familar with how this works...

**Pull request merged**